File: src/agent.ts
#!/usr/bin/env node
import { runGrcAgent } from "./grc-agent.js";

type CliOptions = {
  framework?: string;
  baseline?: string;
  scope?: string;
  interactive?: boolean;
  inputs: string[];
};

type ConvertOptions = {
  to?: string;
  output?: string;
  inputPath?: string;
};

// Manual argument parsing — simple flag-value loop, no external dependency.
function parseArgs(argv: string[]): CliOptions {
  const options: CliOptions = { inputs: [] };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (!arg) continue;
    if (arg === "--framework") {
      options.framework = argv[i + 1];
      i += 1;
      continue;
    }
    if (arg === "--baseline") {
      options.baseline = argv[i + 1];
      i += 1;
      continue;
    }
    if (arg === "--scope") {
      options.scope = argv[i + 1];
      i += 1;
      continue;
    }
    if (arg === "--interactive" || arg === "-i") {
      options.interactive = true;
      continue;
    }
    options.inputs.push(arg);
  }
  return options;
}

function parseConvertArgs(argv: string[]): ConvertOptions {
  const options: ConvertOptions = {};
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (!arg) continue;
    if (arg === "--to") {
      options.to = argv[i + 1];
      i += 1;
      continue;
    }
    if (arg === "--output") {
      options.output = argv[i + 1];
      i += 1;
      continue;
    }
    // First non-flag argument is the input path
    if (!arg.startsWith("--") && !options.inputPath) {
      options.inputPath = arg;
    }
  }
  return options;
}

function usage(): string {
  return [
    "Usage:",
    "  grc-agent --framework 'NIST 800-53' --baseline 'FedRAMP Moderate' --scope 'demo' <paths...>",
    "  grc-agent convert --to oscal-ssp [--output out.json] <input-path>",
    "  grc-agent convert --to oscal-mapping [--output out.json] <input-path>",
    "",
    "Commands:",
    "  convert         Convert an SSP document to OSCAL format",
    "",
    "Options:",
    "  --framework     Framework name (e.g. 'NIST 800-53')",
    "  --baseline      Baseline or level (e.g. 'FedRAMP Moderate')",
    "  --scope         Assessment scope description",
    "  -i, --interactive  Run assessment then enter interactive follow-up mode",
    "",
    "Convert options:",
    "  --to            Target format (oscal-ssp or oscal-mapping)",
    "  --output        Output file path (default: <input-stem>-oscal.json)",
  ].join("\n");
}

async function runConvert(argv: string[]): Promise<void> {
  const opts = parseConvertArgs(argv);
  const validTargets = ["oscal-ssp", "oscal-mapping"];
  if (!opts.to || !validTargets.includes(opts.to)) {
    console.error(`Error: --to must be one of: ${validTargets.join(", ")}\n`);
    console.error(usage());
    process.exit(1);
  }
  if (!opts.inputPath) {
    console.error("Error: input file path is required.\n");
    console.error(usage());
    process.exit(1);
  }

  if (opts.to === "oscal-mapping") {
    const { convertToOscalMapping, defaultMappingOutputPath, writeOscalMapping } = await import(
      "./mapping-convert.js"
    );
    const outputPath = opts.output ?? defaultMappingOutputPath(opts.inputPath);
    console.log(`Converting ${opts.inputPath} to OSCAL mapping-collection format...`);
    const oscalMapping = await convertToOscalMapping(opts.inputPath);
    await writeOscalMapping(oscalMapping, outputPath);
    console.log(`OSCAL mapping-collection written to ${outputPath}`);
    return;
  }

  // Dynamic import keeps OSCAL conversion module out of assessment runs
  const { convertToOscalSsp, defaultOutputPath, writeOscalSsp } = await import(
    "./oscal-convert.js"
  );

  const outputPath = opts.output ?? defaultOutputPath(opts.inputPath);
  console.log(`Converting ${opts.inputPath} to OSCAL SSP format...`);

  const oscalSsp = await convertToOscalSsp(opts.inputPath);
  await writeOscalSsp(oscalSsp, outputPath);

  console.log(`OSCAL SSP written to ${outputPath}`);
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);

  // Route to convert subcommand if first argument is "convert"
  if (argv[0] === "convert") {
    await runConvert(argv.slice(1));
    return;
  }

  const options = parseArgs(argv);
  // All three flags are required: framework, baseline, and scope determine the assessment.
  if (!options.framework || !options.baseline || !options.scope) {
    console.error(usage());
    process.exit(1);
  }

  const input = {
    framework: options.framework,
    baselineOrLevel: options.baseline,
    scope: options.scope,
    inputPaths: options.inputs,
  };

  if (options.interactive) {
    // Dynamic import keeps the readline dependency out of single-shot runs.
    const { runInteractiveSession } = await import("./repl.js");
    await runInteractiveSession(input);
  } else {
    const result = await runGrcAgent(input);
    console.log(JSON.stringify(result, null, 2));
  }
}

main().catch((error) => {
  console.error("Agent failed:", error);
  process.exit(1);
});

File: python/grc_agent/agent.py
#!/usr/bin/env python3
"""CLI entry point for the GRC agent — Python port of src/agent.ts."""

from __future__ import annotations

import argparse
import json
import sys

import asyncio

from .grc_agent import GrcAgentInput, run_grc_agent


# Uses argparse subparsers to route between assessment and convert workflows.
def _parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    args = list(argv or sys.argv[1:])

    if args and args[0] == "convert":
        convert_parser = argparse.ArgumentParser(description="Convert SSP to OSCAL format")
        convert_parser.add_argument(
            "--to",
            required=True,
            choices=["oscal-ssp", "oscal-mapping"],
            help="Target format (oscal-ssp or oscal-mapping)",
        )
        convert_parser.add_argument(
            "--output",
            help="Output file path (default: <input-stem>-oscal.json)",
        )
        convert_parser.add_argument("path", help="Input SSP file path")
        parsed = convert_parser.parse_args(args[1:])
        parsed.command = "convert"  # type: ignore[attr-defined]
        return parsed

    parser = argparse.ArgumentParser(
        description="Multi-framework GRC compliance assessment agent"
    )
    parser.add_argument("--framework", help="Framework name (e.g. 'NIST 800-53')")
    parser.add_argument("--baseline", help="Baseline or level (e.g. 'FedRAMP Moderate')")
    parser.add_argument("--scope", help="Assessment scope description")
    parser.add_argument(
        "-i",
        "--interactive",
        action="store_true",
        help="Run assessment then enter interactive follow-up mode",
    )
    parser.add_argument(
        "paths",
        nargs="*",
        default=[],
        help="Evidence file paths or glob patterns",
    )

    parsed = parser.parse_args(args)
    parsed.command = None  # type: ignore[attr-defined]
    return parsed


async def _run_convert(args: argparse.Namespace) -> None:
    target = getattr(args, "to")

    if target == "oscal-mapping":
        from .mapping_convert import (
            convert_to_oscal_mapping,
            default_mapping_output_path,
            write_oscal_mapping,
        )

        output_path = args.output or default_mapping_output_path(args.path)
        print(f"Converting {args.path} to OSCAL mapping-collection format...")
        oscal_mapping = await convert_to_oscal_mapping(args.path)
        await write_oscal_mapping(oscal_mapping, output_path)
        print(f"OSCAL mapping-collection written to {output_path}")
        return

    from .oscal_convert import convert_to_oscal_ssp, default_output_path, write_oscal_ssp

    output_path = args.output or default_output_path(args.path)
    print(f"Converting {args.path} to OSCAL SSP format...")

    oscal_ssp = await convert_to_oscal_ssp(args.path)
    await write_oscal_ssp(oscal_ssp, output_path)

    print(f"OSCAL SSP written to {output_path}")


async def _async_main(args: argparse.Namespace) -> None:
    # Route to convert subcommand
    if args.command == "convert":
        await _run_convert(args)
        return

    # Assessment mode — framework, baseline, and scope are required
    if not args.framework or not args.baseline or not args.scope:
        print(
            "Error: --framework, --baseline, and --scope are required for assessment mode.",
            file=sys.stderr,
        )
        sys.exit(1)

    inp = GrcAgentInput(
        framework=args.framework,
        baseline_or_level=args.baseline,
        scope=args.scope,
        input_paths=args.paths,
    )

    if args.interactive:
        from .repl import run_interactive_session
        await run_interactive_session(inp)
    else:
        result = await run_grc_agent(inp)
        print(json.dumps(result, indent=2))


def main() -> None:
    args = _parse_args()
    try:
        asyncio.run(_async_main(args))
    except Exception as exc:
        print(f"Agent failed: {exc}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
